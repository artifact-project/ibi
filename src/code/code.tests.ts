import {createDeclarationFactory, createMockFactory} from './code';
import {getRawInterfaces} from 'tx-reflector';

describe('IBox', () => {
	interface IBoxProps {
		color?: string;
		size: number;
		disabled: boolean;
	}

	interface IBoxAttrs {
		color: string;
		size: string;
		disabled: boolean;
	}

	const declarations = createDeclarationFactory<IBoxAttrs>(
		['attrs', 'srcProps'],
		(prop, value, attrs, srcProps) => {
			(srcProps[prop] != null) && (attrs[prop] = srcProps[prop]);
		}
	);

	const boxTx = declarations.create<IBoxProps>('IBox', {
		disabled: true,
		size(prop, value, attrs) {
			attrs[prop] = `${value}px`;
		},
		color(prop, value, attrs, srcProps) {
			attrs[prop] = <string>(srcProps.disabled ? '#ccc' : value);
		}
	});

	it('source', () => {
		expect(boxTx.toString()).toMatchSnapshot();
	});

	it('execute', () => {
		expect(boxTx({}, {
			size: 123,
			disabled: true,
		})).toEqual({
			color: '#ccc',
			size: '123px',
			disabled: true,
		});
	});
});

describe('IBtn', () => {
	interface IBtnProps {
		value: string;
		disabled?: boolean;
	}

	interface IBtnAttrs {
		value: string;
		disabled: boolean;
	}

	const declarations = createDeclarationFactory<IBtnAttrs>(
		['attrs', 'srcProps', 'classNames'],
		(prop, value, attrs, srcProps) => {
			(srcProps[prop] != null) && (attrs[prop] = srcProps[prop]);
		}
	);

	const btnTx = declarations.create<IBtnProps>('IBox', {
		value: true,
		disabled(prop, value, attrs, srcProps, classNames: string[]) {
			attrs[prop] = !!value;
			value && classNames.push(prop);

			if (value) {
				attrs[`aria-${prop}`] = value.toString();
			}
		},
	});

	it('source', () => {
		expect(btnTx.toString()).toMatchSnapshot();
	});

	it('disabled: false', () => {
		const classNames = [];

		expect(btnTx({}, {value: 'Wow!'}, classNames)).toEqual({value: 'Wow!', disabled: false});
		expect(classNames).toEqual([]);
	});

	it('disabled: true', () => {
		const classNames = [];

		expect(btnTx({}, {value: 'Wow!', disabled: true}, classNames)).toEqual({
			value: 'Wow!',
			disabled: true,
			'aria-disabled': 'true',
		});
		expect(classNames).toEqual(['disabled']);
	});
});

it('exec', () => {
	interface IElem {
		tabIndex?: number;
	}

	interface IComponent {
		disabled?: boolean;
	}

	interface IAttr extends IElem, IComponent {
	}

	const declarations = createDeclarationFactory<IAttr>(
		['attrs', 'srcProps'],
		(prop, value, attrs, srcProps) => {
			(srcProps[prop] != null) && (attrs[prop] = srcProps[prop]);
		}
	);

	const elemTx = declarations.create<IElem>('IElem', {
		tabIndex: true,
	});

	const cmpTx = declarations.create<IComponent>('IComponent', {
		disabled: true,
	});

	expect(elemTx.toString()).toMatchSnapshot();
	expect(cmpTx.toString()).toMatchSnapshot();

	const props = {
		tabIndex: 123,
		disabled: false,
	};
	expect(declarations.exec(['IElem', 'IComponent'], {}, props)).toEqual(props);
});

it('mock', () => {
	interface IProps {
		type: string;
		created: Date;
	}

	const mock = createMockFactory<IProps>({
		interfaces: getRawInterfaces<IProps>(),
		propFactory: (entry, parent) => entry.name === 'created' ? '2017-01-02' : parent(entry),
	});

	expect(mock()).toEqual({
		type: '',
		created: '2017-01-02',
	});

	expect(mock({type: 'text'})).toEqual({
		type: 'text',
		created: '2017-01-02',
	});
});
