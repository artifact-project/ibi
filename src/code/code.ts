const R_PARSE_FN = /^function\s*\(([^\)]+)\)\s*\{([\s\S]+)\}/;
const R_ARGS_SEPARATOR = /,\s+/;

export interface TXRule<I, K extends keyof I, A> {
	(prop: K, value: I[K], dst: A, src: I, ...args): void;
}

export interface TXRules<I, A> {
	[propName: string]: true | TXRule<I, keyof I, A>;
}

function jit(source: string, prop: string, args: string[]): string {
	const [, argsStr, body] = source.match(R_PARSE_FN);
	const argsNames = argsStr.trim().split(R_ARGS_SEPARATOR);
	let code = body;

	if (argsNames.length) {
		const propVar = argsNames[0];

		code = code
				.replace(new RegExp(`\\[\\b${propVar}\\b\\]`, 'g'), `.${prop}`)
				.replace(new RegExp(`([\\(\\[ ])${propVar}([ \\]\\)])`, 'g'), `$1'${prop}'$2`)
		;

		if (argsNames.length > 1) {
			const valueVar = argsNames[1];

			code = code.replace(RegExp(`(\\s|[!+-]|\\()(${valueVar})(\\.|\\s|;|\\))`, 'g'), `$1${args[1]}.${prop}$3`);

			if (argsNames.length > 2) {
				argsNames.slice(2).forEach((name, idx) => {
					if (args[idx] !== name) {
						code = `var ${name} = ${args[idx]};\n${code}`;
					}
				});

				// code = code.replace(new RegExp(`\\b${argsNames[2]}\\b(\\.|\\[)`, 'g'), `${args[0]}$1`);
			}
		}
	}

	return `// Inline: ${prop}\n${code}`;
}

export function createDeclarationFactory<A>(args: string[], defaultRule: (prop: string, rule: any, dst: A, src) => void) {
	const entries = {};

	return {
		create<I>(name: string, rules: TXRules<I, A>): (dst: Partial<A>, src: I, ...others) => Partial<A> {
			const source = [];

			Object.keys(rules).forEach(prop => {
				let rule = rules[prop];

				if (rule === true) {
					rule = defaultRule;
				}

				source.push(jit(rule.toString(), prop, args));
			});

			source.push(`return ${args[0]};`);
			entries[name] = Function(`return function ${name}(${args.join(', ')}) {\n${source.join('\n')}\n}`)();

			return entries[name];
		},

		exec(interfaces: string[], dst: Partial<A>, ...others): Partial<A> {
			let idx = interfaces.length;

			while (idx--) {
				const name = interfaces[idx];

				if (entries.hasOwnProperty(name)) {
					entries[name](dst, ...others);
				} else {
					console.warn(`createGenerator().exec: ${name} not register`);
				}
			}

			return dst;
		}
	};
}
