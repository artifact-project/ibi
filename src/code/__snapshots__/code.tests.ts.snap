// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IBox source 1`] = `
"function IBox(attrs, srcProps) {
// Inline: disabled

        srcProps.disabled != null && (attrs.disabled = srcProps.disabled);
    
// Inline: size

            attrs.size = srcProps.size + \\"px\\";
        
// Inline: color

            attrs.color = srcProps.disabled ? '#ccc' : srcProps.color;
        
return attrs;
}"
`;

exports[`IBtn source 1`] = `
"function IBox(attrs, srcProps, classNames) {
// Inline: value

        srcProps.value != null && (attrs.value = srcProps.value);
    
// Inline: disabled

            attrs.disabled = !!srcProps.disabled;
            srcProps.disabled && classNames.push('disabled');
            if (srcProps.disabled) {
                attrs[\\"aria-\\" + 'disabled'] = srcProps.disabled.toString();
            }
        
return attrs;
}"
`;

exports[`exec 1`] = `
"function IElem(attrs, srcProps) {
// Inline: tabIndex

        srcProps.tabIndex != null && (attrs.tabIndex = srcProps.tabIndex);
    
return attrs;
}"
`;

exports[`exec 2`] = `
"function IComponent(attrs, srcProps) {
// Inline: disabled

        srcProps.disabled != null && (attrs.disabled = srcProps.disabled);
    
return attrs;
}"
`;
